File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int reqFreq[26]={0};
        for(char ch: p){
            reqFreq[ch-'a']++;
        }
        vector<int>idx;
        int left=0;
        int currFreq[26]={0};
        for(int right=0;right<s.length();right++){
            currFreq[s[right]-'a']++;
            
            while(currFreq[s[right]-'a']>reqFreq[s[right]-'a']){
                currFreq[s[left]-'a']--;
                left++;
            }
            if(right-left+1==p.length()) idx.push_back(left);

        }
        return idx;
    }
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int freq[]=new int[26];

        for(int i=0; i<p.length();i++){
            freq[p.charAt(i)-'a']++;
        }
        List<Integer>ans=new ArrayList<>();

        int left=0;
        int currFreq[]=new int[26];
        for(int right=0; right<s.length();right++){
            currFreq[s.charAt(right)-'a']++;
            //invalid
            while(currFreq[s.charAt(right)-'a']>freq[s.charAt(right)-'a']){
                currFreq[s.charAt(left)-'a']--;
                left++;
            }
            if(right-left+1==p.length()) ans.add(left);

        }
        return ans;
    }
}